File: FrontEnd/static/js/phaser/utils/eoqDebugLog.js
/**
 * Frontend tracing for Final Shot / FLSS EOQ chains.
 * Filter browser console with: EOQ-TRACE
 *
 * On by default. Disable with: window.GOB_EOQ_TRACE = false
 */

import { pixelsToGrid } from './gridToPixels.js';

export function isEoqTraceEnabled(_scene) {
  if (typeof window !== 'undefined' && window.GOB_EOQ_TRACE === false) {
    return false;
  }
  return true;
}

function roundCoord(n) {
  return Math.round(Number(n) * 100) / 100;
}

export function isEoqTurnData(turnData) {
  if (!turnData || typeof turnData !== 'object') return false;
  return Boolean(
    turnData.final_turn
    || turnData.flss
    || turnData.final_shot_possession
    || turnData.late_clock_eoq
    || turnData.terminal_dreb_eoq
    || turnData.eoq_trace_seq
    || turnData.eoq_trace_role
    || (
      (turnData.result_type === 'BASELINE_INBOUND' || turnData.result_type === 'SIDE_INBOUND')
      && turnData.eoq_trace_seq
    ),
  );
}

export function snapshotClockFromTurn(turnData, scene) {
  const gs = scene?.gameClock?.getState?.() || {};
  return {
    quarter: turnData?.quarter ?? scene?.simData?.quarter ?? scene?.quarter ?? null,
    time_remaining:
      turnData?.time_remaining
      ?? turnData?.clock_end
      ?? gs.remainingSec
      ?? null,
    shot_clock_remaining: turnData?.shot_clock_remaining ?? null,
    clock_start: turnData?.clock_start ?? turnData?.clockStart ?? null,
    clock_end: turnData?.clock_end ?? turnData?.clockEnd ?? null,
    clock_display: turnData?.clock ?? gs.display ?? null,
  };
}

export function snapshotPlayersFromSprites(scene, playerSprites) {
  const width = scene?.game?.config?.width;
  const height = scene?.game?.config?.height;
  const offense = {};
  const defense = {};
  const offenseTeamId = scene?.offenseTeamId ?? scene?.simData?.offense_team_id;

  for (const [playerId, sprite] of Object.entries(playerSprites || {})) {
    if (!sprite) continue;
    const info = scene?.playerInfo?.[playerId];
    const pos = info?.pos ?? '?';
    let coords = null;
    if (Number.isFinite(width) && Number.isFinite(height) && width > 0 && height > 0) {
      coords = pixelsToGrid(sprite.x, sprite.y, width, height);
      coords = { x: roundCoord(coords.x), y: roundCoord(coords.y) };
    } else if (sprite.gridX != null && sprite.gridY != null) {
      coords = { x: roundCoord(sprite.gridX), y: roundCoord(sprite.gridY) };
    }
    const row = {
      player_id: playerId,
      pos,
      coords,
      team_id: sprite.team_id ?? info?.team_id ?? null,
    };
    const isOffense = offenseTeamId != null && String(sprite.team_id) === String(offenseTeamId);
    if (isOffense) {
      offense[pos] = row;
    } else {
      defense[pos] = row;
    }
  }
  return { offense, defense };
}

export function snapshotShooterFromTurn(turnData, scene, playerSprites, label = 'shooter') {
  const shooterId =
    turnData?.shooter_id
    ?? turnData?.shooterId
    ?? turnData?.roles?.shooter_id
    ?? turnData?.ball_handler_id;
  const sprite = shooterId ? playerSprites?.[shooterId] : null;
  const info = shooterId ? scene?.playerInfo?.[shooterId] : null;
  const width = scene?.game?.config?.width;
  const height = scene?.game?.config?.height;
  let coords = turnData?.shooter_coords ?? null;
  if (!coords && sprite && width && height) {
    coords = pixelsToGrid(sprite.x, sprite.y, width, height);
    coords = { x: roundCoord(coords.x), y: roundCoord(coords.y) };
  }
  return {
    label,
    player_id: shooterId ?? null,
    pos: info?.pos ?? turnData?.shooter_pos ?? null,
    name: info?.name ?? null,
    coords,
    flss_zone: turnData?.flss_zone ?? null,
    flss: turnData?.flss ?? false,
    final_turn: turnData?.final_turn ?? false,
  };
}

function clockFromSchemaBlock(block) {
  if (!block || typeof block !== 'object') return {};
  const clock = block.clock && typeof block.clock === 'object' ? block.clock : {};
  return {
    clock_remaining: clock.clock_remaining ?? null,
    shot_clock_remaining: clock.shot_clock_remaining ?? null,
  };
}

export function summarizeAnimationStep(step, index) {
  if (!step || typeof step !== 'object') {
    return { index, error: 'not_an_object' };
  }
  const start = step.start && typeof step.start === 'object' ? step.start : {};
  const end = step.end && typeof step.end === 'object' ? step.end : {};
  const actions = start.action && typeof start.action === 'object' ? start.action : {};
  const actionEntries = Object.entries(actions).slice(0, 8);
  return {
    index,
    id: step.id ?? null,
    start_clock: clockFromSchemaBlock(start),
    end_clock: clockFromSchemaBlock(end),
    duration_ms: step.duration_ms ?? end.time_elapsed ?? null,
    actions: Object.fromEntries(actionEntries),
    next_step_index: step.next_step_index ?? null,
    branch: step.branch ?? null,
  };
}

export function summarizeTurnData(turnData) {
  if (!turnData || typeof turnData !== 'object') return {};
  return {
    eoq_trace_seq: turnData.eoq_trace_seq ?? null,
    eoq_trace_turn_in_seq: turnData.eoq_trace_turn_in_seq ?? null,
    eoq_trace_role: turnData.eoq_trace_role ?? null,
    result_type: turnData.result_type ?? null,
    current_turn: turnData.current_turn ?? null,
    time_elapsed: turnData.time_elapsed ?? null,
    clock_start: turnData.clock_start ?? null,
    clock_end: turnData.clock_end ?? null,
    shot_clock_start: turnData.shot_clock_start ?? null,
    shot_clock_end: turnData.shot_clock_end ?? null,
    next_play_type: turnData.next_play_type ?? null,
    quarter_ends_after: turnData.quarter_ends_after ?? null,
    late_clock_eoq: turnData.late_clock_eoq ?? null,
    flss: turnData.flss ?? null,
    flss_zone: turnData.flss_zone ?? null,
    final_turn: turnData.final_turn ?? null,
    final_shot_possession: turnData.final_shot_possession ?? null,
    terminal_dreb_eoq: turnData.terminal_dreb_eoq ?? null,
    animation_step_count: Array.isArray(turnData.animation_steps)
      ? turnData.animation_steps.length
      : 0,
  };
}

function emitTrace(event, payload) {
  console.warn('[EOQ-TRACE]', event, payload);
}

export function logEoqStep(scene, flow, step, phase, turnData, context = {}, extra = {}) {
  if (!isEoqTraceEnabled(scene)) return;
  const playerSprites = context.playerSprites || scene?.playerSprites || {};
  const payload = {
    event: 'STEP',
    flow,
    step,
    phase,
    turn_index: turnData?.index ?? scene?.currentTurn ?? null,
    turn: summarizeTurnData(turnData),
    clock: snapshotClockFromTurn(turnData, scene),
    players: snapshotPlayersFromSprites(scene, playerSprites),
    shooter: snapshotShooterFromTurn(turnData, scene, playerSprites),
    ...extra,
  };
  emitTrace('STEP', payload);
}

export function logEoqTurn(scene, phase, turnData, context = {}, extra = {}) {
  if (!isEoqTraceEnabled(scene) || !isEoqTurnData(turnData)) return;
  const playerSprites = context.playerSprites || scene?.playerSprites || {};
  const flow = turnData?.flss
    ? 'FLSS'
    : (turnData?.eoq_trace_role || (turnData?.final_turn ? 'FINAL_SHOT' : 'EOQ'));
  const steps = Array.isArray(turnData?.animation_steps) ? turnData.animation_steps : [];
  emitTrace('TURN', {
    event: 'TURN',
    phase,
    flow,
    turn_index: turnData?.index ?? scene?.currentTurn ?? null,
    turn: summarizeTurnData(turnData),
    clock: snapshotClockFromTurn(turnData, scene),
    animation_steps: steps.map((s, i) => summarizeAnimationStep(s, i)),
    players: snapshotPlayersFromSprites(scene, playerSprites),
    shooter: snapshotShooterFromTurn(turnData, scene, playerSprites),
    ...extra,
  });
}

export function logEoqApiReceipt(scene, turnData, responseMeta = {}) {
  if (!isEoqTraceEnabled(scene)) return;
  const turn = turnData?.turn ?? turnData;
  const batch = turn?.result_type === 'BATCH' ? (turn.batch_turns || []) : [turn].filter(Boolean);
  const eoqTurns = batch.filter(isEoqTurnData);
  if (!eoqTurns.length && !responseMeta?.quarter_complete) return;
  emitTrace('API', {
    event: 'API',
    phase: 'RECEIVED',
    response_meta: responseMeta,
    eoq_turns: eoqTurns.map((t) => ({
      turn: summarizeTurnData(t),
      animation_steps: (t.animation_steps || []).map((s, i) => summarizeAnimationStep(s, i)),
    })),
  });
}

export function logEoqSchemaStep(scene, flow, stepIndex, phase, turnData, step, context = {}) {
  if (!isEoqTraceEnabled(scene)) return;
  const playerSprites = context.playerSprites || scene?.playerSprites || {};
  emitTrace('STEP', {
    event: 'STEP',
    flow,
    phase,
    step_index: stepIndex,
    step: summarizeAnimationStep(step, stepIndex),
    turn: summarizeTurnData(turnData),
    clock: snapshotClockFromTurn(turnData, scene),
    players: snapshotPlayersFromSprites(scene, playerSprites),
  });
}

File: BackEnd/engine/eoq_debug_log.py
"""
Structured tracing for Final Shot / FLSS EOQ chains.

Filter server logs with: [EOQ-TRACE]

On by default. Disable with game_state['eoq_trace'] = False (e.g. bulk sims).

Once a Final Shot sequence starts, all related turns in the chain share
``eoq_trace_seq`` (also stamped on turn payloads for frontend correlation):

  FINAL_SHOT → optional BIP/SIP → FLSS → optional BIP → … until quarter end
"""

from __future__ import annotations

import json
import logging
import uuid
from typing import Any, Dict, List, Optional

from BackEnd.utils.shared import get_name_safe, get_player_position

logger = logging.getLogger(__name__)

_TRACE_PREFIX = "[EOQ-TRACE]"


def is_eoq_trace_enabled(game) -> bool:
    gs = getattr(game, "game_state", None) or {}
    return gs.get("eoq_trace") is not False


def is_eoq_turn(result: Any) -> bool:
    if not isinstance(result, dict):
        return False
    if result.get("eoq_trace_role"):
        return True
    return bool(
        result.get("final_turn")
        or result.get("flss")
        or result.get("final_shot_possession")
        or result.get("late_clock_eoq")
        or result.get("terminal_dreb_eoq")
        or result.get("result_type") in ("BASELINE_INBOUND", "SIDE_INBOUND")
        and (
            result.get("late_clock_eoq")
            or (result.get("eoq_trace_seq"))
        )
    )


def clear_eoq_trace_sequence(game) -> None:
    gs = getattr(game, "game_state", None)
    if not isinstance(gs, dict):
        return
    gs.pop("eoq_trace_seq", None)
    gs.pop("eoq_trace_turn_in_seq", None)


def begin_eoq_trace_sequence(game) -> str:
    """Start or continue an EOQ trace sequence; returns sequence id."""
    if not is_eoq_trace_enabled(game):
        return ""
    gs = getattr(game, "game_state", None) or {}
    seq = gs.get("eoq_trace_seq")
    if not seq:
        seq = uuid.uuid4().hex[:10]
        gs["eoq_trace_seq"] = seq
        gs["eoq_trace_turn_in_seq"] = 0
    return str(seq)


def bump_eoq_trace_turn(game) -> int:
    gs = getattr(game, "game_state", None) or {}
    n = int(gs.get("eoq_trace_turn_in_seq") or 0) + 1
    gs["eoq_trace_turn_in_seq"] = n
    return n


def stamp_eoq_trace_on_turn(game, result: Dict[str, Any], role: str) -> None:
    if not isinstance(result, dict) or not is_eoq_trace_enabled(game):
        return
    if result.get("eoq_trace_seq"):
        if role and not result.get("eoq_trace_role"):
            result["eoq_trace_role"] = role
        return
    seq = begin_eoq_trace_sequence(game)
    if not seq:
        return
    turn_in_seq = bump_eoq_trace_turn(game)
    result["eoq_trace_seq"] = seq
    result["eoq_trace_turn_in_seq"] = turn_in_seq
    result["eoq_trace_role"] = role


def snapshot_time(game) -> Dict[str, Any]:
    gs = getattr(game, "game_state", None) or {}
    return {
        "quarter": getattr(game, "quarter", None),
        "time_remaining": gs.get("time_remaining"),
        "shot_clock_remaining": gs.get("shot_clock_remaining"),
        "clock_display": gs.get("clock"),
        "offensive_state": gs.get("offensive_state"),
        "final_shot_possession_active": gs.get("final_shot_possession_active"),
        "final_turn_shot_this_turn": gs.get("final_turn_shot_this_turn"),
    }


def snapshot_chain_state(game) -> Dict[str, Any]:
    gs = getattr(game, "game_state", None) or {}
    return {
        "eoq_trace_seq": gs.get("eoq_trace_seq"),
        "eoq_trace_turn_in_seq": gs.get("eoq_trace_turn_in_seq"),
        "late_clock_eoq_chain_active": gs.get("late_clock_eoq_chain_active"),
        "flss_possession_pending": gs.get("flss_possession_pending"),
        "final_shot_possession_active": gs.get("final_shot_possession_active"),
        "final_turn_shot_this_turn": gs.get("final_turn_shot_this_turn"),
        "free_throws_remaining": gs.get("free_throws_remaining"),
        "pending_oreb": bool(gs.get("pending_oreb")),
    }


def _coords_for_player(player) -> Optional[Dict[str, float]]:
    if player is None:
        return None
    raw = getattr(player, "coords", None) or {}
    if not isinstance(raw, dict):
        return None
    try:
        return {"x": round(float(raw.get("x", 0)), 2), "y": round(float(raw.get("y", 0)), 2)}
    except (TypeError, ValueError):
        return None


def snapshot_all_players(game) -> Dict[str, Any]:
    out: Dict[str, Any] = {"offense": {}, "defense": {}}
    off_team = getattr(game, "offense_team", None)
    def_team = getattr(game, "defense_team", None)
    if off_team and getattr(off_team, "lineup", None):
        for pos, player in off_team.lineup.items():
            if not player:
                continue
            out["offense"][pos] = {
                "player_id": getattr(player, "player_id", None),
                "name": get_name_safe(player),
                "coords": _coords_for_player(player),
            }
    if def_team and getattr(def_team, "lineup", None):
        for pos, player in def_team.lineup.items():
            if not player:
                continue
            out["defense"][pos] = {
                "player_id": getattr(player, "player_id", None),
                "name": get_name_safe(player),
                "coords": _coords_for_player(player),
            }
    return out


def snapshot_shooter(
    game,
    player,
    *,
    pos: Optional[str] = None,
    label: str = "shooter",
) -> Optional[Dict[str, Any]]:
    if player is None:
        return None
    if pos is None:
        lineup = getattr(game.offense_team, "lineup", None) or {}
        pos = get_player_position(lineup, player)
    return {
        "label": label,
        "pos": pos,
        "player_id": getattr(player, "player_id", None),
        "name": get_name_safe(player),
        "coords": _coords_for_player(player),
    }


def _clock_from_schema_block(block: Any) -> Dict[str, Any]:
    if not isinstance(block, dict):
        return {}
    clock = block.get("clock") if isinstance(block.get("clock"), dict) else {}
    return {
        "clock_remaining": clock.get("clock_remaining"),
        "shot_clock_remaining": clock.get("shot_clock_remaining"),
    }


def summarize_animation_step(step: Any, index: int) -> Dict[str, Any]:
    if not isinstance(step, dict):
        return {"index": index, "error": "not_a_dict"}
    start = step.get("start") if isinstance(step.get("start"), dict) else {}
    end = step.get("end") if isinstance(step.get("end"), dict) else {}
    actions = start.get("action") if isinstance(start.get("action"), dict) else {}
    action_summary = {
        pid: act for pid, act in list(actions.items())[:8]
    }
    return {
        "index": index,
        "id": step.get("id"),
        "start_clock": _clock_from_schema_block(start),
        "end_clock": _clock_from_schema_block(end),
        "duration_ms": step.get("duration_ms") or end.get("time_elapsed"),
        "actions": action_summary,
        "next_step_index": step.get("next_step_index"),
        "branch": step.get("branch"),
        "metadata": step.get("metadata") if isinstance(step.get("metadata"), dict) else None,
    }


def summarize_animation_steps(steps: Any) -> List[Dict[str, Any]]:
    if not isinstance(steps, list):
        return []
    return [summarize_animation_step(step, i) for i, step in enumerate(steps)]


def summarize_turn_result(result: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "eoq_trace_seq": result.get("eoq_trace_seq"),
        "eoq_trace_turn_in_seq": result.get("eoq_trace_turn_in_seq"),
        "eoq_trace_role": result.get("eoq_trace_role"),
        "result_type": result.get("result_type"),
        "current_turn": result.get("current_turn"),
        "text": (result.get("text") or "")[:120],
        "time_elapsed": result.get("time_elapsed"),
        "clock_start": result.get("clock_start"),
        "clock_end": result.get("clock_end"),
        "shot_clock_start": result.get("shot_clock_start"),
        "shot_clock_end": result.get("shot_clock_end"),
        "real_time_elapsed_ms": result.get("real_time_elapsed_ms"),
        "next_play_type": result.get("next_play_type"),
        "next_turn": result.get("next_turn"),
        "quarter_ends_after": result.get("quarter_ends_after"),
        "late_clock_eoq": result.get("late_clock_eoq"),
        "terminal_dreb_eoq": result.get("terminal_dreb_eoq"),
        "flss": result.get("flss"),
        "flss_zone": result.get("flss_zone"),
        "final_turn": result.get("final_turn"),
        "final_shot_possession": result.get("final_shot_possession"),
        "forced_shot_reason": result.get("forced_shot_reason"),
        "final_turn_anchor_clock": result.get("final_turn_anchor_clock"),
        "shooter_id": result.get("shooter_id"),
        "possession_flips": result.get("possession_flips"),
        "animation_step_count": len(result.get("animation_steps") or []),
        "clock_contract_source": result.get("clock_contract_source"),
    }


def _emit_trace(event: str, payload: Dict[str, Any]) -> None:
    try:
        body = json.dumps(payload, default=str)
    except TypeError:
        body = str(payload)
    logger.warning("%s %s %s", _TRACE_PREFIX, event, body)


def log_eoq_step(
    game,
    flow: str,
    step: str,
    phase: str,
    *,
    shooter=None,
    shooter_pos: Optional[str] = None,
    extra: Optional[Dict[str, Any]] = None,
) -> None:
    if not is_eoq_trace_enabled(game):
        return
    payload: Dict[str, Any] = {
        "event": "STEP",
        "flow": flow,
        "step": step,
        "phase": phase,
        "chain": snapshot_chain_state(game),
        "clock": snapshot_time(game),
        "players": snapshot_all_players(game),
    }
    if shooter is not None or shooter_pos is not None:
        payload["shooter"] = snapshot_shooter(game, shooter, pos=shooter_pos)
    if extra:
        payload["extra"] = extra
    _emit_trace("STEP", payload)


def log_eoq_chain_event(
    game,
    event: str,
    *,
    extra: Optional[Dict[str, Any]] = None,
) -> None:
    if not is_eoq_trace_enabled(game):
        return
    payload: Dict[str, Any] = {
        "event": "CHAIN",
        "chain_event": event,
        "chain": snapshot_chain_state(game),
        "clock": snapshot_time(game),
    }
    if extra:
        payload["extra"] = extra
    _emit_trace("CHAIN", payload)


def log_eoq_turn(
    game,
    role: str,
    result: Dict[str, Any],
    *,
    phase: str,
    turn_num: Optional[int] = None,
    game_id: Optional[str] = None,
    extra: Optional[Dict[str, Any]] = None,
) -> None:
    """
    Full turn bundle: flags, clock contract, animation step summaries.

    role: FINAL_SHOT | FLSS | BIP | SIP | DREB_TERMINAL | OREB | FREE_THROW
    phase: PRE_EMIT | POST_EMIT | POST_CLOCK | API_RESPONSE
    """
    if not is_eoq_trace_enabled(game):
        return
    if phase in ("PRE_EMIT", "POST_EMIT", "POST_CLOCK"):
        stamp_eoq_trace_on_turn(game, result, role)
    steps = result.get("animation_steps") or []
    payload: Dict[str, Any] = {
        "event": "TURN",
        "phase": phase,
        "role": role,
        "turn_num": turn_num,
        "game_id": game_id or getattr(game, "game_id", None),
        "chain": snapshot_chain_state(game),
        "clock": snapshot_time(game),
        "turn": summarize_turn_result(result),
        "animation_steps": summarize_animation_steps(steps),
        "players": snapshot_all_players(game),
    }
    if extra:
        payload["extra"] = extra
    _emit_trace("TURN", payload)


def log_eoq_api_response(
    game,
    *,
    game_id: Optional[str],
    turn_num: Optional[int],
    turns_in_response: List[Dict[str, Any]],
    response_meta: Dict[str, Any],
) -> None:
    if not is_eoq_trace_enabled(game):
        return
    eoq_turns = []
    for i, t in enumerate(turns_in_response or []):
        if not isinstance(t, dict):
            continue
        if not (
            t.get("final_turn")
            or t.get("flss")
            or t.get("late_clock_eoq")
            or t.get("final_shot_possession")
            or t.get("eoq_trace_role")
            or t.get("terminal_dreb_eoq")
            or (
                t.get("result_type") in ("BASELINE_INBOUND", "SIDE_INBOUND")
                and t.get("eoq_trace_seq")
            )
        ):
            continue
        eoq_turns.append(
            {
                "batch_index": i,
                "turn": summarize_turn_result(t),
                "animation_steps": summarize_animation_steps(t.get("animation_steps") or []),
            }
        )
    if not eoq_turns and not snapshot_chain_state(game).get("eoq_trace_seq"):
        return
    payload = {
        "event": "API",
        "game_id": game_id,
        "turn_num": turn_num,
        "chain": snapshot_chain_state(game),
        "clock": snapshot_time(game),
        "response_meta": response_meta,
        "eoq_turns_in_response": eoq_turns,
    }
    _emit_trace("API", payload)


def log_eoq_routing_decision(
    game,
    *,
    branch: str,
    game_clock_remaining,
    would_final_shot: bool,
    extra: Optional[Dict[str, Any]] = None,
) -> None:
    if not is_eoq_trace_enabled(game):
        return
    log_eoq_step(
        game,
        "ROUTING",
        branch,
        "DECISION",
        extra={
            "game_clock_remaining": game_clock_remaining,
            "would_take_final_shot": would_final_shot,
            **(extra or {}),
        },
    )
